spectator entity location based on eye level (WIP)

bradcarnage opened the pull request:
(static "cat" string needs to be swapped out for either via config load or some other method)

bradcarnage committed:
spectator entity location based on eye level

(static "cat" string needs to be swapped out for either via config load or some other method)

rfresh2 commented on src/main/java/com/zenith/network/server/handler/spectator/incoming/movement/PlayerPositionRotationSpectatorHandler.java:
```diff
@@ -29,31 +31,47 @@ public static void updateSpectatorPosition(final ServerConnection selfSession) {
             return;
         }
+        double playerEyeHeight = 1.62;
+        EntityStats entityStats = new EntityStats();
```
Integrate any additional metadata about spectator entity types in the SpectatorEntity class hierarchy under here: https://github.com/rfresh2/ZenithProxy/tree/1.20/src/main/java/com/zenith/feature/spectator

The current spectator entity type can be accessed here on the ServerConnection: https://github.com/rfresh2/ZenithProxy/blob/c04b583187cbb0dc5d745fd4bb5ec562f804cae6/src/main/java/com/zenith/network/server/ServerConnection.java#L78

so you shouldn't need any `EntityData` class initializations or hardcoded map lookups in the `updateSpectatorPosition` function.

I'm imagining we add a set of methods on SpectatorEntity to get whatever data is needed about the current type.

bradcarnage replied:
change was made

rfresh2 commented on src/main/java/com/zenith/util/math/MathHelper.java:
```diff
@@ -44,4 +44,20 @@ public static float wrapDegrees(float degrees) {
         return Math.max(min, Math.min(max, value));
     }
+
+    public static double[] translateEntity(double x, double y, double z, float yaw, float pitch, double distance) {
+        double yawRadians = Math.toRadians(yaw);
+        double pitchRadians = Math.toRadians(pitch);
+
+        double xOffset = -Math.sin(yawRadians) * Math.cos(pitchRadians) * distance;
+        double yOffset = -Math.sin(pitchRadians) * distance;
+        double zOffset = Math.cos(yawRadians) * Math.cos(pitchRadians) * distance;
+
+        double newX = x + xOffset;
+        double newY = y + yOffset;
+        double newZ = z + zOffset;
+
+        return new double[]{newX,newY,newZ};
+    }
```
I'd prefer inlining this to where its needed so we avoid heap allocations on the hotpath of packet processing. 

I don't think this func is being reused or shared anywhere else

bradcarnage replied:
change was made

bradcarnage committed:
removed EntityStats.java
moved eyeHeight, totalHeight, totalWidth to SpectatorEntity classes
send adjusted position to main player as well

bradcarnage committed:
remove unused mathhelper function

rfresh2 committed:
cleanup and fix a couple edge cases

rfresh2 commented:
Thanks!
